Bluetooth: SSP: Support non-bondable mode

lylezhu2012 opened the pull request:
Add a function bt_get_bondable to get the bonding setting.

If the return value of function bt_get_bondable is false, clear the bonding flag when controller requiring `Authentication_Requirements`.

lylezhu2012 committed:
Bluetooth: Host: smp: Add function to get bonding setting

Add a function bt_get_bondable to get the bonding setting.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: SSP: Support non-bondable mode

If the return value of function bt_get_bondable is false, clear the
bonding flag when controller requiring `Authentication_Requirements`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented on subsys/bluetooth/host/smp.c:
```diff
@@ -2639,6 +2639,11 @@ void bt_set_bondable(bool enable)
 	bondable = enable;
 }
 
+bool bt_get_bondable(void)
+{
+	return bondable;
+}
```
Does `ssp` use `smp`? 

If so, then `latch_bondable` should be used instead of this as our API supports setting `bondable` per connection.

lylezhu2012 replied:
No. `ssp` does not use `smp`.

lylezhu2012 replied:
Actually, `sm->bondable` is copy of `bondable`.
```
static bool latch_bondable(struct bt_smp *smp)
{
	(void)atomic_cas(&smp->bondable, BT_SMP_BONDABLE_UNINITIALIZED, (atomic_val_t)bondable);

	return atomic_get(&smp->bondable);
}
```

My change is also get a copy of `bondable`.

Thalley replied:
> Actually, sm->bondable is copy of bondable.

`smp->bondable` can be overridden by `bt_conn_set_bondable`, so it's not always the same value as `bondable`

alwa-nordic replied:
> Then why are returning a `bondable` flag from SMP?

Arguably, the bondable setting does not belong in SMP. "Bondable mode" is a GAP thing.

So ideally, there should be a function named `bt_gap_set_bondable`.

A lower-level API that bypasses GAP, it should be named `bt_smp_set_bondable`. Not sure if it's a good idea to expose such a function though.

There should be no function with the non-specific name `bt_set_bondable` to minimize confusion.

Thalley replied:
> Arguably, the bondable setting does not belong in SMP. "Bondable mode" is a GAP thing.

Then I guess the `bondable` value should also be placed outside of SMP (`smp.c`) (e.g. in `conn.c`)?

alwa-nordic replied:
Putting it in `conn.c` is OK, but I would prefer `gap.c`. That aligns better with how the core spec is organized.

lylezhu2012 replied:
> > Actually, sm->bondable is copy of bondable.
> 
> `smp->bondable` can be overridden by `bt_conn_set_bondable`, so it's not always the same value as `bondable`

`ssp` has not a object likes as `smp`. For current implementation, function `bt_conn_set_bondable` is unsupported by BR/EDR. But the function `bt_conn_set_bondable` should be supported by BR/EDR. I will update PR to support it.

lylezhu2012 replied:
> > Then why are returning a `bondable` flag from SMP?
> 
> Arguably, the bondable setting does not belong in SMP. "Bondable mode" is a GAP thing.
> 
> So ideally, the function would be named `bt_gap_set_bondable`. There should be no function with the non-specific name `bt_set_bondable` to minimize confusion.
> 
> Lower-level APIs that switch between SMP taking the value from GAP and a different source should be prefixed with `bt_smp_`.



> > Arguably, the bondable setting does not belong in SMP. "Bondable mode" is a GAP thing.
> 
> Then I guess the `bondable` value should also be placed outside of SMP (`smp.c`) (e.g. in `conn.c`)?



> Putting it in `conn.c` is OK, but I would prefer `gap.c`. That aligns better with how the core spec is organized.

I recommend making a separate PR for this change. I think we can create a task for it.

Thalley replied:
> Putting it in `conn.c` is OK, but I would prefer `gap.c`. That aligns better with how the core spec is organized.

We don't have a `gap.c` yet, but I agree. A lot of what's inside of `conn.c` arguably belongs in `gap.c` (and ditto for `conn.h` and `gap.h`), but that'll be a larger change :)

alwa-nordic commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -504,6 +504,9 @@ void bt_finalize_init(void);
 
 void bt_hci_host_num_completed_packets(struct net_buf *buf);
 
+/* Get bondable flag */
+bool bt_get_bondable(void);
+
```
I would like this to be a public function, like `bt_set_bondable` is.

lylezhu2012 replied:
Currently, I haven’t found any need for this outside of stack, so I made it an internal function.
Public it is OK for me.

Thalley replied:
Agree with @alwa-nordic - Since setting `bondable` is a public function, getting it might as well be that too.

lylezhu2012 committed:
Bluetooth: BR: Improve `bt_conn_set_bondable`

In current, the bondable flag cannot be configured for each specific
BR connection.
But for LE conn, there is a function `bt_conn_set_bondable` for this
purpose.

Improve `bt_conn_set_bondable` to support BR conn.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1426,6 +1426,16 @@ static inline const char *bt_security_err_to_str(enum bt_security_err err)
+ *
+ *  Get current Bonding flag.
+ *  The initial value of this flag depends on BT_BONDABLE Kconfig setting.
+ *  And the Bonding flag can be updated using bt_set_bondable.
```
```suggestion
 *  The Bonding flag can be updated using bt_set_bondable().
```

lylezhu2012 replied:
Updated.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1426,6 +1426,16 @@ static inline const char *bt_security_err_to_str(enum bt_security_err err)
+/** @brief Get bonding flag.
+ *
+ *  Get current Bonding flag.
+ *  The initial value of this flag depends on BT_BONDABLE Kconfig setting.
```
```suggestion
 *  The initial value of this flag depends on @kconfig{CONFIG_BT_BONDABLE} setting.
```

lylezhu2012 replied:
Updated.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1426,6 +1426,16 @@ static inline const char *bt_security_err_to_str(enum bt_security_err err)
 
+/** @brief Get bonding flag.
+ *
+ *  Get current Bonding flag.
```
```suggestion
 *  Get current bonding flag.
```

lylezhu2012 replied:
Updated.

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -1426,6 +1426,16 @@ static inline const char *bt_security_err_to_str(enum bt_security_err err)
+ *  The initial value of this flag depends on BT_BONDABLE Kconfig setting.
+ *  And the Bonding flag can be updated using bt_set_bondable.
+ *
+ *  @return Current Bonding flag.
```
```suggestion
 *  @return Current bonding flag.
```

lylezhu2012 replied:
Updated.

lylezhu2012 committed:
Bluetooth: shell: Add command `conn-bondable`

Add command `conn-bondable` to enable/disable the pairing bondable
flag of a specific ACL connection.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: SSP: Incorrect bonding status notified

The `bonded` flag of the callback `pairing_complete` is always true,
event if the SSP pairing is non-bondable.

Check the bonding status in SSP complete event instead of in link key
notify event.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented on subsys/bluetooth/shell/bredr.c:
```diff
@@ -531,6 +531,41 @@ static int cmd_sdp_find_record(const struct shell *sh,
 	return 0;
 }
 
+#if defined(CONFIG_BT_BONDABLE_PER_CONNECTION)
+static int cmd_conn_bondable(const struct shell *sh, size_t argc, char *argv[])
+{
+	int res;
+	const char *action;
+	bool enable;
+
+	if (!default_conn) {
+		shell_print(sh, "Not connected");
+		return 0;
+	}
+
+	action = argv[1];
+
+	if (!strcmp(action, "on")) {
+		enable = true;
+	} else if (!strcmp(action, "off")) {
+		enable = false;
+	} else {
+		shell_help(sh);
+		return 0;
+	}
+
+	shell_print(sh, "[%p] set conn bondable %s", default_conn, action);
+
+	res = bt_conn_set_bondable(default_conn, enable);
+	if (res) {
+		shell_error(sh, "Set conn bondable failed: result %d", res);
+		return -ENOEXEC;
+	}
+	shell_print(sh, "Set conn bondable done");
+	return 0;
+}
+#endif /* CONFIG_BT_BONDABLE_PER_CONNECTION */
```
This isn't BR/EDR specific; should be in the `bt.c` file instead

lylezhu2012 replied:
Done.

Thalley commented on subsys/bluetooth/host/smp.c:
```diff
@@ -5416,6 +5421,14 @@ int bt_conn_set_bondable(struct bt_conn *conn, bool enable)
 	struct bt_smp *smp;
 
+	if (IS_ENABLED(CONFIG_BT_CLASSIC) && (conn->type == BT_CONN_TYPE_BR)) {
+		if (atomic_test_bit(conn->flags, BT_CONN_BR_BONDABLE) == enable) {
+			return -EALREADY;
+		}
+		atomic_set_bit_to(conn->flags, BT_CONN_BR_BONDABLE, enable);
+		return 0;
```
This isn't thread safe

Suggest
```suggestion
		if (enable && atomic_test_and_set_bit(conn->flags, BT_CONN_BR_BONDABLE)) {
			return -EALREADY;
		} else if (!enable && atomic_test_and_clear_bit(conn->flags, BT_CONN_BR_BONDABLE)) {
			return -EALREADY;
		}
		
		return 0;
```

lylezhu2012 replied:
Updated.

Thalley commented on subsys/bluetooth/host/conn_internal.h:
If we keep adding additional conn flags, we might want to start guarding them more, as each added flag requires an additional bit, so there might be some easy optimization to be found here

Thalley commented on subsys/bluetooth/host/smp.c:
```diff
@@ -5416,6 +5421,16 @@ int bt_conn_set_bondable(struct bt_conn *conn, bool enable)
 {
 	struct bt_smp *smp;
 
+	if (IS_ENABLED(CONFIG_BT_CLASSIC) && (conn->type == BT_CONN_TYPE_BR)) {
```
```suggestion
	if (IS_ENABLED(CONFIG_BT_CLASSIC) && IS_ENABLED(CONFIG_BT_BONDABLE) && (conn->type == BT_CONN_TYPE_BR)) {
```

lylezhu2012 replied:
For the implementation and comments of function `bt_conn_set_bondable`, I think the `CONFIG_BT_BONDABLE` only affects the initialisation value.

Thalley replied:
`CONFIG_BT_BONDABLE` enables the entire bonding feature in LE:

```
#if defined(CONFIG_BT_BONDABLE)
#define BT_SMP_AUTH_BONDING_FLAGS BT_SMP_AUTH_BONDING
#else
#define BT_SMP_AUTH_BONDING_FLAGS 0
#endif /* CONFIG_BT_BONDABLE */
```

I think there are probably some functions (like `bt_set_bondable`) that should be guarded with this, as if `CONFIG_BT_BONDABLE=n` then the `bondable` value is effectively not used. 

In that regard it does not make sense to perform any of the checks here if it's disabled.

lylezhu2012 replied:
I think it is not at all. When `CONFIG_BT_BONDABLE=n` and `bt_conn_set_bondable(conn, true)` called, the `auth_req` of pairing request or security request will set bit `BT_SMP_AUTH_BONDING`.
Please refer to https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/smp.c#L2889 or https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/smp.c#L3135.

https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/smp.c#L2670-L2674

Thalley replied:
@jhedberg You are also more than welcome to weigh in here. Does it looks like there's an issue with `BT_BONDABLE` if it can be set to 0 but it's still possible to be bondable via `bt_conn_set_bondable`?

jhedberg replied:
@Thalley the way I see it, is that the Kconfig option is about supporting bondable mode, whereas the runtime API is about enabling/disabling it.

jhedberg replied:
> @Thalley the way I see it, is that the Kconfig option is about supporting bondable mode, whereas the runtime API is about enabling/disabling it.

If we currently allow toggling bondable even with the Kconfig disabled, then this interpretation doesn't seem correct (and the Kconfig text is misleading)

jhedberg replied:
We have this:
```c
static bool bondable = IS_ENABLED(CONFIG_BT_BONDABLE);
```

So the meaning of the current Kconfig option is just about setting the default state of `bondable`. This is not reflected in the Kconfig help text.

Another issue I see in `host/smp.c` is stuff like this, which seems to completely ignore any runtime state:
```c
#if defined(CONFIG_BT_BONDABLE)
#define BT_SMP_AUTH_BONDING_FLAGS BT_SMP_AUTH_BONDING
#else
#define BT_SMP_AUTH_BONDING_FLAGS 0
#endif /* CONFIG_BT_BONDABLE */
```

Thalley replied:
Right, so the implementation does not match the description of the Kconfig option. So we have to adjust either of them

jhedberg replied:
After reading through the `smp.c` code a bit more it seems like the Kconfig help text should probably be updated. The bonding flags macros also don't seem to have much effect, rather the code will overwrite those both ways depending on the runtime setting. Anyway, all of that seems to be out of scope for this PR, so I'll approve here.

Thalley replied:
Indeed, it is outside the scope of this PR, the question is mostly whether this PR should consider `CONFIG_BT_BONDABLE` per the text or per the implementation.

jhedberg replied:
> Indeed, it is outside the scope of this PR, the question is mostly whether this PR should consider `CONFIG_BT_BONDABLE` per the text or per the implementation.

Sorry, I missed the detail where this actually has different behavior than the existing implementation, i.e. uses Kconfig to guard against the entry into this BR/EDR branch. @lylezhu2012 could you fix that?

jhedberg replied:
> @lylezhu2012 could you fix that?

FWIW, feel free to add a second commit here to fix the Kconfig help text as well, if you want (otherwise it'll need to be fixed in a follow-up PR).

lylezhu2012 replied:
Yes. I have added a commit to update it.

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -3739,6 +3739,41 @@ static int cmd_bondable(const struct shell *sh, size_t argc, char *argv[])
+		return 0;
+	}
+
+	action = argv[1];
+
+	if (!strcmp(action, "on")) {
+		enable = true;
+	} else if (!strcmp(action, "off")) {
+		enable = false;
+	} else {
+		shell_help(sh);
+		return 0;
+	}
```
Suggest to use `shell_strtobool` which handles multiple ways of setting boolean values and has error checking

Thalley replied:
> Or I need to give a initialized value before passing the parameter `err`. Interesting.

Yes, they have been designed to require the `err` being initialized to 0. I don't recall the reason why :)

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -3739,6 +3739,41 @@ static int cmd_bondable(const struct shell *sh, size_t argc, char *argv[])
+	} else if (!strcmp(action, "off")) {
+		enable = false;
+	} else {
+		shell_help(sh);
+		return 0;
```
```suggestion
		shell_help(sh);
		return SHELL_CMD_HELP_PRINTED;
```

lylezhu2012 replied:
Updated.

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -3739,6 +3739,41 @@ static int cmd_bondable(const struct shell *sh, size_t argc, char *argv[])
+#if defined(CONFIG_BT_BONDABLE_PER_CONNECTION)
+static int cmd_conn_bondable(const struct shell *sh, size_t argc, char *argv[])
+{
+	int res;
```
```suggestion
	int err;
```

It's more common to use `err` rather than `res` in this file

lylezhu2012 replied:
Updated. However, it seems that the previous code did not follow this rule.

Thalley replied:
What previous code?

lylezhu2012 replied:
I mean `int res` is referred from other shell command.

Thalley replied:
`res` is not used within `bt.c` - Might be in the classic shell module :)

lylezhu2012 replied:
You are right. It seams `bredr.c` has not be maintained. I will create a PR this issue.

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -3739,6 +3739,41 @@ static int cmd_bondable(const struct shell *sh, size_t argc, char *argv[])
+
+	if (!default_conn) {
+		shell_print(sh, "Not connected");
+		return 0;
```
```suggestion
		return -ENOEXEC;
```

lylezhu2012 replied:
Updated.

jhedberg commented:
@lylezhu2012 minor cosmetic requests: second commit message shouldn't call it "binding" (it's a different word than bonding and not used in a bluetooth context). IIRC it's formally called "Bonding" and "Non-bondable mode" in the specification. Also, your commit messages are again almost one word per line, so please try to fix the line width in your editor settings.

lylezhu2012 commented:
> on-bondable mode

Yes. Updated all commits.

lylezhu2012 committed:
Bluetooth: host: Kconfig: Correct `help` of `BT_BONDABLE`

Correct the `help` of the configuration `BT_BONDABLE`.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

Thalley commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -632,9 +632,14 @@ config BT_BONDABLE
-	  This option enables support for Bondable Mode. In this mode,
-	  Bonding flag in AuthReq of SMP Pairing Request/Response will be set
-	  indicating the support for this mode.
+	  This option enables support for Bondable Mode. This option is the
+	  default value of the bonding flag for any ACL connection. If the option
```
```suggestion
	  This option is the
	  default value of the bonding flag for any ACL connection. If the option
```

Isn't this more correct? As far as I can tell, a device can still enter the bondable mode even if this is disabled, so this option does not control the bondable mode, but only what the initial bondable state is. 

Arguably, since the bondable state can be changed at runtime, we should just remove this option. If we reduce it to just control the initial value of `bondable`, then it's pretty useless alongside `bt_set_bondable`. 

If we actually want to control whether bonding is support or not by this Kconfig, that's apparently a larger change.

lylezhu2012 replied:
Yes, But if this config is removed, how to control he initial value of `bondable`?

> Arguably, since the bondable state can be changed at runtime, we should just remove this option. If we reduce it to just control the initial value of `bondable`, then it's pretty useless alongside `bt_set_bondable`.

jhedberg replied:
The sane default is true, as per the Kconfig default. That said, this is a breaking change which needs to be documented. Luckily, removing the option will cause a build failure for any out of tree app that so far tried to disable it, so it won't go silently unnoticed.

dleach02 commented:
@lylezhu2012 have you resolved the change request from @alwa-nordic?
